### PR TITLE
Port and refactor robust access tests

### PR DESCRIPTION
**Pull request by kainino0x**
Commits can be viewed separately (though you may want to view commits 3/4/5 together).

Passes in Chrome Canary except for ~https://crbug.com/tint/953 and~ https://crbug.com/tint/938

Replaces #380

<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

**Comment by kainino0x** [comment]
@Kangz PTAL

File: src/webgpu/shader/execution/robust_access.spec.ts [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+  atomic?: boolean;
+}) {
+  const scalarInfo = kScalarTypeInfo[baseType];
+  if (atomic) assert(scalarInfo.supportsAtomics, 'type does not support atomics');
+  const scalarType = atomic ? `atomic<${baseType}>` : baseType;
+
+  // Vector types
+  for (const vectorType of kVectorContainerTypes) {
+    yield {
+      type: `${vectorType}<${scalarType}>`,
```
Comment (Kangz) [inline src/webgpu/shader/execution/robust_access.spec.ts:136]: Are you allowed to do `vec4<atomic<f32>>`??

File: src/webgpu/shader/execution/robust_access.spec.ts [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+      })
+      .filter(p => {
+        const info = kScalarTypeInfo[p.baseType];
+
+        // Storage and uniform require host-sharable types.
+        const isHostShared = p.storageClass === 'storage' || p.storageClass === 'uniform';
+        if (isHostShared && info.layout === undefined) return false;
+
+        // Atomic access requires atomic type and storage/workgroup memory.
+        const classSupportsAtomics = p.storageClass === 'storage' || p.storageClass === 'workgroup';
```
Comment (Kangz) [inline src/webgpu/shader/execution/robust_access.spec.ts:214]: Maybe this check could be in the `.combine('atomic')` that would instead become an expand.

File: src/webgpu/shader/execution/robust_access.spec.ts [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+      let errorReturnValue = 0x1000;
+      return () => {
+        ++errorReturnValue;
+        return `0x${errorReturnValue.toString(16)}u`;
+      };
+    })();
+
+    // Produce the accesses to the variable.
+    for (const indexToTest of indicesToTest) {
+      // TODO check with constants too if WGSL allows it.
```
Comment (Kangz) [inline src/webgpu/shader/execution/robust_access.spec.ts:327]: I think it meant constexpr but I'm not sure. Could be overridable constants too, that would make sense as well.

File: src/webgpu/shader/execution/robust_access.spec.ts [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+      // FIXME: ^ What does this comment mean, does it mean pipeline-overridable constants?
+      //          Move this TODO into the test description once that's figured out.
+      const index = `(${indexToTest}) + ${constZeroIndex}`;
+      const zeroElement = `${_typeInfo.elementType}()`;
+
+      switch (access) {
+        case 'read':
+          {
+            let element = `s.data[${index}]`;
```
Comment (Kangz) [inline src/webgpu/shader/execution/robust_access.spec.ts:336]: I think this text can be shared between read and write.

File: src/webgpu/shader/execution/robust_access.spec.ts [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+      const testBuffer = t.device.createBuffer({
+        mappedAtCreation: true,
+        size: testBufferSize,
+        usage:
+          GPUBufferUsage.COPY_SRC |
+          GPUBufferUsage.UNIFORM |
+          GPUBufferUsage.STORAGE |
+          GPUBufferUsage.COPY_DST,
+      });
+      memcpy({ src: expectedData, dst: testBuffer.getMappedRange() });
```
Comment (Kangz) [inline src/webgpu/shader/execution/robust_access.spec.ts:402]: Maybe we can call testFillArrayBuffer directly on getMappedRange?

**Review by Kangz — APPROVED** [review]
LGTM

File: src/webgpu/shader/execution/robust_access.spec.ts (hunk 1 above)
Comment (kainino0x) [reply src/webgpu/shader/execution/robust_access.spec.ts:136]: I'm not sure! Tint doesn't stop me, at least. I figure we'll revise the test if this turns out to be not allowed.

Comment (kainino0x) [reply src/webgpu/shader/execution/robust_access.spec.ts:136]: I found that Tint also crashes in the same way on a clearly wrong thing, `vec2<vec2<>>`, so I'm going to assume it's not supposed to be allowed and take it out.

File: src/webgpu/shader/execution/robust_access.spec.ts (hunk 5 above)
Comment (kainino0x) [reply src/webgpu/shader/execution/robust_access.spec.ts:402]: I realized I could just use `t.makeBufferWithContents(expectedData, usage)`

**Pull request merged**